File: LeetCode - Algorithm/Dynamic Programming/300.py
# 300. Longest Increasing Subsequence
# https://leetcode.com/problems/longest-increasing-subsequence/

from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        memo = [1] * len(nums)

        current = 1

        while current < len(memo):
            for i in range(current):
                if nums[current] > nums[i]:
                    memo[current] = max(memo[current], memo[i] + 1)

            current += 1

        return max(memo)

File: LeetCode - Algorithm/Dynamic Programming/300.js
// 300. Longest Increasing Subsequence
// https://leetcode.com/problems/longest-increasing-subsequence/

/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function (nums) {
  let result = 1;

  const memo = [];
  for (let i = 0; i < nums.length; i++) memo.push(1);

  for (let i = 1; i < nums.length; i++) {
    for (let j = 0; j < i; j++) {
      if (nums[i] > nums[j]) {
        memo[i] = Math.max(memo[i], memo[j] + 1);
        result = Math.max(result, memo[i]);
      }
    }
  }

  return result;
};

/*
nums = [10,9,2,5,3,7,101,18]
         ^ ^
*/
